When the federation function is enabled, an exception occurs when tpcc is executed.

### Which version of ShardingSphere did you use?
we find java version: java8, full_version=1.8.0_342, full_path=/home/peilq_sharding/bisheng-jdk1.8.0_342//bin/java
ShardingSphere-5.2.2-SNAPSHOT
Commit ID: dirty-753c0cee8ee6fd3db00536da55b64bc5198a3758
Commit Message: Optimize sqlFederationExecutor init logic when sqlFederationType modify dynamically (https://github.com/apache/shardingsphere/pull/22209)
Branch: https://github.com/apache/shardingsphere/commit/753c0cee8ee6fd3db00536da55b64bc5198a3758
Build time: 2022-11-19T10:18:41+0800

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy

### Expected behavior
When the federation function is enabled, tpcc is executed successfully.
### Actual behavior
When the federation function is enabled, an exception occurs when tpcc is executed.

```
12:23:11,457 [main] INFO   jTPCC : Term-00,                                                                                          12:23:13,754 [Thread-5] ERROR  jTPCCTData : Unexpected SQLException in STOCK_LEVELge: 308MB / 1963MB
12:23:13,754 [Thread-5] ERROR  jTPCCTData : [90.90.44.173:32424/90.90.44.175:11000] ERROR: no vertex rel#3096:HepRelVertex(rel#3092:LogicalFilter.NONE.[](input=HepRelVertex#3058,condition=AND(=($0, ?0), <($2, ?1))))
org.opengauss.util.PSQLException: [90.90.44.173:32424/90.90.44.175:11000] ERROR: no vertex rel#3096:HepRelVertex(rel#3092:LogicalFilter.NONE.[](input=HepRelVertex#3058,condition=AND(=($0, ?0), <($2, ?1))))
        at org.opengauss.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2901)
        at org.opengauss.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2630)
        at org.opengauss.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:362)
        at org.opengauss.jdbc.PgStatement.runQueryExecutor(PgStatement.java:562)
        at org.opengauss.jdbc.PgStatement.executeInternal(PgStatement.java:539)
        at org.opengauss.jdbc.PgStatement.execute(PgStatement.java:397)
        at org.opengauss.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:156)
        at org.opengauss.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:112)
        at jTPCCTData.executeStockLevel(jTPCCTData.java:1279)
        at jTPCCTData.execute(jTPCCTData.java:101)
        at jTPCCTerminal.executeTransactions(jTPCCTerminal.java:172)
        at jTPCCTerminal.run(jTPCCTerminal.java:87)
        at java.lang.Thread.run(Thread.java:748)
12:23:13,769 [Thread-9] ERROR  jTPCCTData : Unexpected SQLException in STOCK_LEVEL
12:23:13,769 [Thread-9] ERROR  jTPCCTData : [90.90.44.173:32492/90.90.44.175:11000] ERROR: java.lang.AssertionError: Relational expression rel#3088:LogicalAggregate.NONE.[](input=LogicalJoin#3097,group={},low_stock=COUNT()) belongs to a different planner than is currently being used.
org.opengauss.util.PSQLException: [90.90.44.173:32492/90.90.44.175:11000] ERROR: java.lang.AssertionError: Relational expression rel#3088:LogicalAggregate.NONE.[](input=LogicalJoin#3097,group={},low_stock=COUNT()) belongs to a different planner than is currently being used.
        at org.opengauss.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2901)
        at org.opengauss.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2630)
        at org.opengauss.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:362)
        at org.opengauss.jdbc.PgStatement.runQueryExecutor(PgStatement.java:562)
        at org.opengauss.jdbc.PgStatement.executeInternal(PgStatement.java:539)
        at org.opengauss.jdbc.PgStatement.execute(PgStatement.java:397)
        at org.opengauss.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:156)
        at org.opengauss.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:112)
        at jTPCCTData.executeStockLevel(jTPCCTData.java:1279)
        at jTPCCTData.execute(jTPCCTData.java:101)
        at jTPCCTerminal.executeTransactions(jTPCCTerminal.java:172)
        at jTPCCTerminal.run(jTPCCTerminal.java:87)
        at java.lang.Thread.run(Thread.java:748)    
```


```
14:21:44,699 [main] INFO   jTPCC : Term-00, C value for C_LAST this run:    113
14:21:44,699 [main] INFO   jTPCC : Term-00,                                                                                         14:        21:47,801 [Thread-1] ERROR  jTPCCTData : Unexpected SQLException in STOCK_LEVELage: 318MB / 1963MB
14:21:47,801 [Thread-4] ERROR  jTPCCTData : Unexpected SQLException in STOCK_LEVEL
14:21:47,801 [Thread-2] ERROR  jTPCCTData : Unexpected SQLException in STOCK_LEVEL
14:21:47,802 [Thread-4] ERROR  jTPCCTData : [90.90.44.173:53566/90.90.44.175:11000] ERROR: org.apache.calcite.rel.metadata.CyclicMetada        taException
14:21:47,802 [Thread-1] ERROR  jTPCCTData : [90.90.44.173:53500/90.90.44.175:11000] ERROR: org.apache.calcite.rel.metadata.CyclicMetada        taException
14:21:47,802 [Thread-2] ERROR  jTPCCTData : [90.90.44.173:53530/90.90.44.175:11000] ERROR: Index: 0, Size: 1
org.opengauss.util.PSQLException: [90.90.44.173:53566/90.90.44.175:11000] ERROR: org.apache.calcite.rel.metadata.CyclicMetadataExceptio        n
        at org.opengauss.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2901)
        at org.opengauss.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2630)
        at org.opengauss.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:362)
        at org.opengauss.jdbc.PgStatement.runQueryExecutor(PgStatement.java:562)
        at org.opengauss.jdbc.PgStatement.executeInternal(PgStatement.java:539)
        at org.opengauss.jdbc.PgStatement.execute(PgStatement.java:397)
        at org.opengauss.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:156)
        at org.opengauss.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:112)
        at jTPCCTData.executeStockLevel(jTPCCTData.java:1279)
        at jTPCCTData.execute(jTPCCTData.java:101)
        at jTPCCTerminal.executeTransactions(jTPCCTerminal.java:172)
        at jTPCCTerminal.run(jTPCCTerminal.java:87)
        at java.lang.Thread.run(Thread.java:748)
org.opengauss.util.PSQLException: [90.90.44.173:53500/90.90.44.175:11000] ERROR: org.apache.calcite.rel.metadata.CyclicMetadataExceptio        n
        at org.opengauss.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2901)
        at org.opengauss.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2630)
        at org.opengauss.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:362)
        at org.opengauss.jdbc.PgStatement.runQueryExecutor(PgStatement.java:562)
        at org.opengauss.jdbc.PgStatement.executeInternal(PgStatement.java:539)
        at org.opengauss.jdbc.PgStatement.execute(PgStatement.java:397)
        at org.opengauss.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:156)
        at org.opengauss.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:112)
        at jTPCCTData.executeStockLevel(jTPCCTData.java:1279)
        at jTPCCTData.execute(jTPCCTData.java:101)
        at jTPCCTerminal.executeTransactions(jTPCCTerminal.java:172)
        at jTPCCTerminal.run(jTPCCTerminal.java:87)
        at java.lang.Thread.run(Thread.java:748)
org.opengauss.util.PSQLException: [90.90.44.173:53530/90.90.44.175:11000] ERROR: Index: 0, Size: 1
        at org.opengauss.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2901)
        at org.opengauss.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2630)
        at org.opengauss.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:362)
        at org.opengauss.jdbc.PgStatement.runQueryExecutor(PgStatement.java:562)
        at org.opengauss.jdbc.PgStatement.executeInternal(PgStatement.java:539)
        at org.opengauss.jdbc.PgStatement.execute(PgStatement.java:397)
        at org.opengauss.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:156)
        at org.opengauss.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:112)
        at jTPCCTData.executeStockLevel(jTPCCTData.java:1279)
        at jTPCCTData.execute(jTPCCTData.java:101)
        at jTPCCTerminal.executeTransactions(jTPCCTerminal.java:172)
        at jTPCCTerminal.run(jTPCCTerminal.java:87)
        at java.lang.Thread.run(Thread.java:748)
14:21:47,838 [Thread-5] ERROR  jTPCCTData : Unexpected SQLException in STOCK_LEVEL
14:21:47,839 [Thread-5] ERROR  jTPCCTData : [90.90.44.173:53586/90.90.44.175:11000] ERROR: java.lang.AssertionError: Relational express        ion rel#2628:LogicalAggregate.NONE.[](input=LogicalJoin#2653,group={},low_stock=COUNT()) belongs to a different planner than is current        ly being used.
org.opengauss.util.PSQLException: [90.90.44.173:53586/90.90.44.175:11000] ERROR: java.lang.AssertionError: Relational expression rel#26        28:LogicalAggregate.NONE.[](input=LogicalJoin#2653,group={},low_stock=COUNT()) belongs to a different planner than is currently being u        sed.
        at org.opengauss.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2901)
        at org.opengauss.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2630)
        at org.opengauss.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:362)
        at org.opengauss.jdbc.PgStatement.runQueryExecutor(PgStatement.java:562)
        at org.opengauss.jdbc.PgStatement.executeInternal(PgStatement.java:539)
        at org.opengauss.jdbc.PgStatement.execute(PgStatement.java:397)
        at org.opengauss.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:156)
        at org.opengauss.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:112)
        at jTPCCTData.executeStockLevel(jTPCCTData.java:1279)
        at jTPCCTData.execute(jTPCCTData.java:101)
        at jTPCCTerminal.executeTransactions(jTPCCTerminal.java:172)
        at jTPCCTerminal.run(jTPCCTerminal.java:87)
        at java.lang.Thread.run(Thread.java:748)                                        
```      
### Reason analyze (If you can)

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
gsql -d tpccdb -p 11000 -h 90.90.44.175 -U sharding -W sharding -r
gsql ((openGauss 3.0.0 build cad66d01) compiled at 2022-08-31 20:15:45 commit 0 last mr  )
Non-SSL connection (SSL connection is recommended when requiring high-security)
Type "help" for help.

tpccdb=> set dist variable sql_federation_type="NONE";
SUCCESS
tpccdb=> set dist variable sql_federation_type="ADVANCED";
SUCCESS

step1：set dist variable sql_federation_type="NONE";     and then “. runBenchmark.sh props.proxy”   --successfully

step2：set dist variable sql_federation_type="ADVANCED";     and then “. runBenchmark.sh props.proxy”  --failure


### Example codes for reproduce this issue (such as a github link).
```
schemaName: tpcc_db
dataSources:
  ds_0:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 260
    minPoolSize: 10
    password: Test@123
    url: jdbc:opengauss://90.90.44.171:14000/tpccdb?batchMode=on
    username: tpccuser
  ds_1:
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 260
    minPoolSize: 10
    password: Test@123
    url: jdbc:opengauss://90.90.44.171:15000/tpccdb?batchMode=on
    username: tpccuser

rules:
- !SHARDING
  bindingTables:
  - bmsql_warehouse, bmsql_customer
  - bmsql_stock, bmsql_district, bmsql_order_line
  defaultDatabaseStrategy:
    standard:
      shardingAlgorithmName: database_inline
      shardingColumn: ds_id
  defaultTableStrategy:
    none: null
  shardingAlgorithms:
    database_inline:
      props:
        algorithm-expression: ds_${ds_id % 1}
      type: INLINE
    ds_bmsql_item_inline:
      props:
        algorithm-expression: ds_${i_id % 2}
      type: INLINE
    ds_bmsql_customer_inline:
      props:
        algorithm-expression: ds_${c_w_id % 2}
      type: INLINE
    ds_bmsql_district_inline:
      props:
        algorithm-expression: ds_${d_w_id % 2}
      type: INLINE
    ds_bmsql_history_inline:
      props:
        algorithm-expression: ds_${h_w_id % 2}
      type: INLINE
    ds_bmsql_new_order_inline:
      props:
        algorithm-expression: ds_${no_w_id % 2}
      type: INLINE
    ds_bmsql_oorder_inline:
      props:
        algorithm-expression: ds_${o_w_id % 2}
      type: INLINE
    ds_bmsql_order_line_inline:
      props:
        algorithm-expression: ds_${ol_w_id % 2}
      type: INLINE
    ds_bmsql_stock_inline:
      props:
        algorithm-expression: ds_${s_w_id % 2}
      type: INLINE
    ds_bmsql_warehouse_inline:
      props:
        algorithm-expression: ds_${w_id % 2}
      type: INLINE
  tables:
    bmsql_config:
      actualDataNodes: ds_0.bmsql_config
    bmsql_item:
      actualDataNodes: ds_${0..1}.bmsql_item
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_item_inline
          shardingColumn: i_id
    bmsql_customer:
      actualDataNodes: ds_${0..1}.bmsql_customer
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_customer_inline
          shardingColumn: c_w_id
    bmsql_district:
      actualDataNodes: ds_${0..1}.bmsql_district
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_district_inline
          shardingColumn: d_w_id
    bmsql_history:
      actualDataNodes: ds_${0..1}.bmsql_history
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_history_inline
          shardingColumn: h_w_id
    bmsql_new_order:
      actualDataNodes: ds_${0..1}.bmsql_new_order
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_new_order_inline
          shardingColumn: no_w_id
    bmsql_oorder:
      actualDataNodes: ds_${0..1}.bmsql_oorder
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_oorder_inline
          shardingColumn: o_w_id
    bmsql_order_line:
      actualDataNodes: ds_${0..1}.bmsql_order_line
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_order_line_inline
          shardingColumn: ol_w_id
    bmsql_stock:
      actualDataNodes: ds_${0..1}.bmsql_stock
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_stock_inline
          shardingColumn: s_w_id
    bmsql_warehouse:
      actualDataNodes: ds_${0..1}.bmsql_warehouse
      databaseStrategy:
        standard:
          shardingAlgorithmName: ds_bmsql_warehouse_inline
          shardingColumn: w_id
```

```
db=postgres
driver=org.opengauss.Driver
conn=jdbc:opengauss://90.90.44.175:11000/tpccdb?loggerLevel=OFF
user=sharding
password=sharding

warehouses=5
loadWorkers=10

terminals=20
//To run specified transactions per terminal- runMins must equal zero
runTxnsPerTerminal=0
//To run for specified minutes- runTxnsPerTerminal must equal zero
runMins=2
//Number of total transactions per minute
limitTxnsPerMin=0

//Set to true to run in 4.x compatible mode. Set to false to use the
//entire configured database evenly.
terminalWarehouseFixed=false

//The following five values must add up to 100
//The default percentages of 45, 43, 4, 4 & 4 match the TPC-C spec
newOrderWeight=45
paymentWeight=43
orderStatusWeight=4
deliveryWeight=4
stockLevelWeight=4

// Directory name to create for collecting detailed result data.
// Comment this out to suppress.
//resultDirectory=my_result_%tY-%tm-%td_%tH%tM%tS
//osCollectorScript=./misc/os_collector_linux.py
//osCollectorInterval=1
//osCollectorSSHAddr=user@dbhost
//osCollectorDevices=net_eth0 blk_sda

```

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/decider/ShardingSQLFederationDecider.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/decider/ShardingSQLFederationDecider.java
@@ -48,7 +48,7 @@ public final class ShardingSQLFederationDecider implements SQLFederationDecider<
         }
         addTableDataNodes(deciderContext, rule, tableNames);
         ShardingConditions shardingConditions = getMergedShardingConditions(queryContext, database, rule);
-        if (shardingConditions.isSameShardingCondition()) {
+        if (shardingConditions.isNeedMerge() && shardingConditions.isSameShardingCondition()) {
             return;
         }
         if (select.isContainsSubquery() || select.isContainsHaving() || select.isContainsCombine() || select.isContainsPartialDistinctAggregation()) {

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/decider/ShardingSQLFederationDecider.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/decider/ShardingSQLFederationDecider.java
@@ -47,8 +47,8 @@ public final class ShardingSQLFederationDecider implements SQLFederationDecider<
             return;
         }
         addTableDataNodes(deciderContext, rule, tableNames);
-        ShardingConditions shardingConditions = createShardingConditions(queryContext, database, rule);
-        if (shardingConditions.isNeedMerge() && shardingConditions.isSameShardingCondition()) {
+        ShardingConditions shardingConditions = getMergedShardingConditions(queryContext, database, rule);
+        if (shardingConditions.isSameShardingCondition()) {
             return;
         }
         if (select.isContainsSubquery() || select.isContainsHaving() || select.isContainsCombine() || select.isContainsPartialDistinctAggregation()) {
@@ -69,10 +69,14 @@ public final class ShardingSQLFederationDecider implements SQLFederationDecider<
     }
     
     @SuppressWarnings({"unchecked", "rawtypes"})
-    private static ShardingConditions createShardingConditions(final QueryContext queryContext, final ShardingSphereDatabase database, final ShardingRule rule) {
+    private static ShardingConditions getMergedShardingConditions(final QueryContext queryContext, final ShardingSphereDatabase database, final ShardingRule rule) {
         ShardingConditionEngine shardingConditionEngine = ShardingConditionEngineFactory.createShardingConditionEngine(queryContext, database, rule);
         List<ShardingCondition> shardingConditions = shardingConditionEngine.createShardingConditions(queryContext.getSqlStatementContext(), queryContext.getParameters());
-        return new ShardingConditions(shardingConditions, queryContext.getSqlStatementContext(), rule);
+        ShardingConditions result = new ShardingConditions(shardingConditions, queryContext.getSqlStatementContext(), rule);
+        if (result.isNeedMerge()) {
+            result.merge();
+        }
+        return result;
     }
     
     @Override

--- a/features/sharding/core/src/test/java/org/apache/shardingsphere/sharding/decider/ShardingSQLFederationDeciderTest.java
+++ b/features/sharding/core/src/test/java/org/apache/shardingsphere/sharding/decider/ShardingSQLFederationDeciderTest.java
@@ -25,12 +25,20 @@ import org.apache.shardingsphere.infra.database.DefaultDatabase;
 import org.apache.shardingsphere.infra.database.type.dialect.PostgreSQLDatabaseType;
 import org.apache.shardingsphere.infra.datanode.DataNode;
 import org.apache.shardingsphere.infra.metadata.database.ShardingSphereDatabase;
+import org.apache.shardingsphere.sharding.route.engine.condition.ShardingCondition;
+import org.apache.shardingsphere.sharding.route.engine.condition.engine.ShardingConditionEngineFactory;
+import org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.WhereClauseShardingConditionEngine;
+import org.apache.shardingsphere.sharding.route.engine.condition.value.ListShardingConditionValue;
+import org.apache.shardingsphere.sharding.rule.BindingTableRule;
 import org.apache.shardingsphere.sharding.rule.ShardingRule;
 import org.apache.shardingsphere.sharding.rule.TableRule;
 import org.junit.Test;
+import org.mockito.MockedStatic;
 
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.List;
 import java.util.Optional;
 import java.util.Properties;
 
@@ -38,8 +46,10 @@ import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
 import static org.mockito.Mockito.when;
 
 public final class ShardingSQLFederationDeciderTest {
@@ -58,15 +68,31 @@ public final class ShardingSQLFederationDeciderTest {
     }
     
     @Test
-    public void assertDecideWhenContainsPagination() {
+    public void assertDecideWhenContainsSameShardingCondition() {
         SelectStatementContext select = createStatementContext();
-        when(select.getPaginationContext().isHasPagination()).thenReturn(true);
+        when(select.isContainsSubquery()).thenReturn(true);
         QueryContext queryContext = new QueryContext(select, "", Collections.emptyList());
         SQLFederationDeciderContext actual = new SQLFederationDeciderContext();
         ShardingSQLFederationDecider federationDecider = new ShardingSQLFederationDecider();
-        federationDecider.decide(actual, queryContext, createDatabase(), createShardingRule(), new ConfigurationProperties(new Properties()));
+        try (MockedStatic<ShardingConditionEngineFactory> shardingConditionEngineFactory = mockStatic(ShardingConditionEngineFactory.class)) {
+            WhereClauseShardingConditionEngine shardingConditionEngine = mock(WhereClauseShardingConditionEngine.class);
+            when(shardingConditionEngine.createShardingConditions(any(), any())).thenReturn(createShardingConditions());
+            shardingConditionEngineFactory.when(() -> ShardingConditionEngineFactory.createShardingConditionEngine(any(QueryContext.class), any(), any())).thenReturn(shardingConditionEngine);
+            federationDecider.decide(actual, queryContext, createDatabase(), createShardingRule(), new ConfigurationProperties(new Properties()));
+        }
         assertThat(actual.getDataNodes().size(), is(4));
         assertFalse(actual.isUseSQLFederation());
+    }
+    
+    private List<ShardingCondition> createShardingConditions() {
+        List<ShardingCondition> result = new ArrayList<>();
+        ShardingCondition shardingCondition1 = new ShardingCondition();
+        shardingCondition1.getValues().add(new ListShardingConditionValue<>("order_id", "t_order", Collections.singletonList(1)));
+        result.add(shardingCondition1);
+        ShardingCondition shardingCondition2 = new ShardingCondition();
+        shardingCondition2.getValues().add(new ListShardingConditionValue<>("order_id", "t_order_item", Collections.singletonList(1)));
+        result.add(shardingCondition2);
+        return result;
     }
     
     @Test
@@ -198,6 +224,11 @@ public final class ShardingSQLFederationDeciderTest {
                 new DataNode("ds_0", "t_order"), new DataNode("ds_1", "t_order"),
                 new DataNode("ds_0", "t_order_item"), new DataNode("ds_1", "t_order_item")));
         when(result.findTableRule("t_order")).thenReturn(Optional.of(tableRule));
+        BindingTableRule bindingTableRule = mock(BindingTableRule.class);
+        when(bindingTableRule.hasLogicTable("t_order")).thenReturn(true);
+        when(bindingTableRule.hasLogicTable("t_order_item")).thenReturn(true);
+        when(result.findBindingTableRule("t_order")).thenReturn(Optional.of(bindingTableRule));
+        when(result.findBindingTableRule("t_order_item")).thenReturn(Optional.of(bindingTableRule));
         return result;
     }
 }

